core/src/main/java/org/geysermc/geyser/entity/type/LivingEntity.java

```diff
@@ -278,6 +283,29 @@ public void setParticles(ObjectEntityMetadata<List<Particle>> entityMetadata) {
         }
     }
 
+    public boolean isOnClimbableBlock() {
+        Vector3i blockPos = this.position.toInt();
+        BlockState state = session.getGeyser().getWorldManager().blockAt(session, blockPos);
+        if (session.getTagCache().is(BlockTag.CLIMBABLE, state.block())) {
+            return true;
+        }
+
+        if (state.block() instanceof TrapDoorBlock) {
+            if (!state.getValue(Properties.OPEN)) {
+                return false;
+            } else {
+                BlockState belowState = session.getGeyser().getWorldManager().blockAt(session, blockPos.down());
+                return belowState.is(Blocks.LADDER) && belowState.getValue(Properties.HORIZONTAL_FACING) == state.getValue(Properties.HORIZONTAL_FACING);
+            }
+        }
+        return false;
+    }
```

Comment:
SessionPlayerEntity#isOnClimbableBlock could probably also call this - just to reduce code duplication